trio/monitor.py

```diff
@@ -0,0 +1,380 @@
+
+            elif key == "before_io_wait":
+                timeout = item[1]
+                message = "Waiting for IO (timeout: {:.3f})".format(timeout)
+
+            elif key == "after_io_wait":
+                timeout = item[1]
+                message = "Done waiting for IO (timeout: {:.3f})"\
+                    .format(timeout)
+
+            elif key == "before_task_step":
+                task = item[1]
+                message = "Task stepping: {} ({})".format(task.name, id(task))
+
+            elif key == "after_task_step":
+                task = item[1]
+                message = "Task finished stepping: {} ({})".format(
+                    task.name, id(task)
+                )
+
+            else:
+                message = "Unknown event: {}".format(key)
+
+            message = prefix + message
+
+            try:
+                await stream.send_all(message.encode("ascii") + b'\n')
+            except BrokenStreamError:  # client disconnected on us
+                return
+
+    # command definitions
+    async def command_help(self):
+        """Sends help."""
+        name_rpad = 12
+
+        def pred(i):
+            return hasattr(i, "__name__") \
+                   and i.__name__.startswith("command_")
+
+        commands = inspect.getmembers(self, predicate=pred)
+        lines = ["Commands:"]
+        for name, command in commands:
+            doc = inspect.getdoc(command).splitlines(keepends=False)[0]
+            name = name.split("_", 1)[1]
+            lines.append(name.ljust(name_rpad) + doc)
+
+        return lines
+
+    async def command_signal(self, signame: str):
+        """Sends a signal to the server process."""
+        signame = signame.upper()
+        if not signame.startswith("SIG"):
+            signame = "SIG{}".format(signame)
+
+        try:
+            tosend = getattr(signal, signame)
+        except AttributeError:
+            return ["Invalid signal: {}".format(signame)]
+
+        os.kill(os.getpid(), tosend)
+        return ["Signal sent successfully"]
+
+    async def command_ps(self):
+        """Gets the current list of tasks."""
+        lines = []
+        headers = ('ID', 'Name', 'Parent')
+        widths = (15, 50, 15)
+        header_line = []
+
+        for name, width in zip(headers, widths):
+            header_line.append(name.ljust(width))
+
+        lines.append(' '.join(header_line))
+        lines.append("-" * sum(widths))
+
+        for task in self.flatten_tasks():
+            if len(task.name) >= 50:
+                name = task.name[:46] + "..."
+            else:
+                name = task.name
+
+            if task.parent_nursery is None:
+                parent = "N/A"
+            else:
+                parent = str(id(task.parent_nursery.parent_task))
+
+            result = [str(id(task)).ljust(widths[0]), name.ljust(49), parent]
+
+            lines.append(' '.join(result))
+
+        return lines
```

Comment:
I wonder if we should be showing a hierarchical task tree view as much as possible, as like... a matter of principle almost. People coming to trio are used to unstructured global task lists in other systems (Unix processes, threads, asyncio, ...), and it's just not the right image to have in your head when you're trying to reason about a trio program, so reinforcing that seems like it might be important?

Also, as a usability thing, it'd probably be helpful to have a single command that dumps a combined task tree + stack traces of all tasks, for the use case where you're trying to do remote debugging via a confused user ("can you run the monitor, type <XX>, and paste the output you get into this issue?"). Sort of like how `thread apply all bt` is a magic incantation you ask people to type into gdb.